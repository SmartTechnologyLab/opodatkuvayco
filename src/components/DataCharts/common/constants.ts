export const stackedGraphOptions = {
  maintainAspectRatio: false,
  aspectRatio: 0.8,
  plugins: {
    tooltips: {
      mode: 'index',
      intersect: false
    }
  },
  scales: {
    x: {
      stacked: true
    },
    y: {
      stacked: true
    }
  }
}
