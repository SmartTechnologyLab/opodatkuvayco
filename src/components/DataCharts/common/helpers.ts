import type { Deal, Operation } from '@/components/common/DataTable/mocks'
import type { ChartData } from '@/components/common/UiChart/common/types'
import { getMonthesMap, getMonthNameByNumber } from '@/lib/dateUtils'
import { pipe, curry } from 'ramda'

const groupPurchaseByMonth = curry((operation: Operation, deals: Deal[]) => {
  return deals.reduce(
    (acc, deal) => {
      const month = getMonthNameByNumber(deal[operation].date)

      if (!acc[month]) {
        acc[month] = []
      }

      acc[month].push(deal[operation].uah)

      return acc
    },
    {} as Record<string, number[]>
  )
})

const groupDealsWithMonthMap = curry((operation: Operation, deals: Deal[]) => {
  return deals.reduce(
    (acc, deal) => {
      const monthesMap = getMonthesMap()
      const month = getMonthNameByNumber(deal[operation].date)

      if (!acc[deal.ticker]) {
        acc[deal.ticker] = monthesMap
      }

      acc[deal.ticker][month].push(deal)

      return acc
    },
    {} as Record<string, Record<string, Deal[]>>
  )
})

const groupPurchaseByTicker = (deals: Deal[]): Record<string, Deal[]> => {
  return deals.reduce(
    (acc, deal) => {
      if (!acc[deal.ticker]) {
        acc[deal.ticker] = []
      }

      acc[deal.ticker].push(deal)

      return acc
    },
    {} as Record<string, Deal[]>
  )
}

const sumDealsPropertyByEachMonth = (
  dealsMap: Record<string, Record<string, Deal[]>>
): Record<string, Record<string, number>> => {
  const result: Record<string, Record<string, number>> = {}

  for (const key in dealsMap) {
    for (const month in dealsMap[key]) {
      if (!result[key]) {
        result[key] = {}
      }

      if (!dealsMap[key][month].length) {
        result[key][month] = 0
      } else {
        result[key][month] = dealsMap[key][month].reduce((acc, deal) => acc + deal.purchase.uah, 0)
      }
    }
  }

  return result
}

const getFilteredMonthsWithTrades = (dealsByMonthMap: Record<string, Record<string, number>>) => {
  const monthes: string[] = []

  for (const key in dealsByMonthMap) {
    for (const month in dealsByMonthMap[key]) {
      if (!!dealsByMonthMap[key][month] && !monthes.includes(month)) {
        monthes.push(month)
      }
    }
  }

  return monthes
}

const resultArrayOfDealsByEachMonth = (dealsMap: Record<string, Record<string, number>>): ChartData => {
  const dataSet = []

  for (const key in dealsMap) {
    const data = []

    for (const month in dealsMap[key]) {
      data.push(dealsMap[key][month])
    }

    dataSet.push({
      label: key,
      data
    })
  }

  return {
    labels: getFilteredMonthsWithTrades(dealsMap),
    datasets: dataSet
  }
}

const getAllTickersValue = (deals: Deal[]) => {
  return deals.reduce((acc, deal) => {
    if (!acc.includes(deal.ticker)) {
      acc.push(deal.ticker)
    }

    return acc
  }, [] as string[])
}

const arraySum = (arr: Record<string, number[]>): Record<string, number> => {
  const resultMap: Record<string, number> = {}

  for (const key in arr) {
    resultMap[key] = arr[key].reduce((acc, curr) => acc + curr, 0)
  }

  return resultMap
}

const sumDealValues = (dealsMap: Record<string, Deal[]>): Record<string, number> => {
  const result: Record<string, number> = {}

  for (const key in dealsMap) {
    result[key] = dealsMap[key].reduce((acc, value) => acc + value.quantity, 0)
  }

  return result
}

const getAverageValue = (groupedDeals: Record<string, number>): Record<string, number> => {
  const totalValue = Object.values(groupedDeals).reduce((acc, curr) => acc + curr, 0)

  const dealsAverageMap: Record<string, number> = {}

  for (const key in groupedDeals) {
    dealsAverageMap[key] = +(groupedDeals[key] / totalValue).toFixed(2) * 100
  }

  return dealsAverageMap
}

const getTradesPercentageResultByTicker = pipe(groupPurchaseByTicker, sumDealValues, getAverageValue)

const getResultByTickerEachMonth = (operation: Operation) =>
  pipe(groupDealsWithMonthMap(operation), sumDealsPropertyByEachMonth, resultArrayOfDealsByEachMonth)

const getRevenueByPurchaseMonth = (operation: Operation) => pipe(groupPurchaseByMonth(operation), arraySum)

export { getTradesPercentageResultByTicker, getRevenueByPurchaseMonth, getAllTickersValue, getResultByTickerEachMonth }
