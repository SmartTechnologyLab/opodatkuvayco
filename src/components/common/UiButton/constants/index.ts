export enum Icons {
  CHECK = 'pi pi-check',
  USER = 'pi pi-user',
  CROSS = 'pi pi-times',
  DELETE = 'pi pi-trash',
  PLUS = 'pi pi-plus',
  EDIT = 'pi pi-pencil',
  IMPORT = 'pi pi-file-arrow-up'
}

export enum Severities {
  SECONDARY = 'secondary',
  HELP = 'help',
  DANGER = 'danger'
}
