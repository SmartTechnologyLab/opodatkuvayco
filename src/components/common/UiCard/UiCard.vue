<script setup lang="ts">
import Card from 'primevue/card'
</script>

<template>
  <Card class="card">
    <template #title>
      <slot name="title" />
    </template>
    <template #content>
      <slot />
    </template>
  </Card>
</template>

<style scoped lang="scss">
$card-background: $main-text-color;

.card {
  max-width: 760px;
  background: $card-background;
  color: black;
}
</style>
