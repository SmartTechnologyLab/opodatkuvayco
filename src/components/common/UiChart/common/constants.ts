const documentStyle = getComputedStyle(document.body)

export const CSS_VARIABLES = [
  documentStyle.getPropertyValue('--blue-200'),
  documentStyle.getPropertyValue('--green-200'),
  documentStyle.getPropertyValue('--yellow-200')
]

export const DEFAULT_CHART_OPTIONS = {
  maintainAspectRatio: false
}
