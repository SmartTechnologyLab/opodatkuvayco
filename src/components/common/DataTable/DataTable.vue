<script setup lang="ts">
import DataTable, { type DataTableCellEditCompleteEvent } from 'primevue/datatable'
import Column from 'primevue/column'
import Card from 'primevue/card'
import { path } from 'ramda'
import { useI18n } from 'vue-i18n'
import { type Ref } from 'vue'
import { type Deal } from '@/components/common/DataTable/mocks'
import { FormatType, type Table } from '@/components/common/DataTable/types'
import { isNil } from 'ramda'
import UiInput from '@/components/common/UiInput/UiInput.vue'
import Calendar from 'primevue/calendar'
import { calendarFields, currencyFields, numberFields } from '@/components/common/DataTable/constants/fieldsList'
import UiNumberInput from '../UiNumberInput/UiNumberInput.vue'

defineProps<{
  table: Table
  title?: string
  sortableColumn?: boolean
  removeSortable?: boolean
  notEditableColumns?: string[]
}>()

defineEmits<{
  (e: 'onCellEditComplete', event: DataTableCellEditCompleteEvent): void
}>()

const { d, n } = useI18n()

const getFormattedData = (data: Ref<Deal[]>, field: string, type?: FormatType) => {
  const value = path(field.split('.'), data)

  const [formatted] = [
    type === FormatType.Date && d(value),
    type === FormatType.Number && n(value),
    type === FormatType.CurrencyUSD && n(value, { style: 'currency', currency: 'USD' }),
    type === FormatType.CurrencyUAH && n(value, 'currency'),
    type === FormatType.Percent && n(value, 'percent'),
    value
  ].filter(Boolean)

  return formatted
}

const isFieldEditable = (notEditableColumns: string[] | undefined, field: string) => {
  return !notEditableColumns?.includes(field)
}
</script>

<template>
  <Card>
    <template #title v-if="title">
      {{ $t(title) }}
    </template>

    <template #content>
      <DataTable
        :value="table.data"
        size="small"
        :removableSort="removeSortable"
        @cell-edit-complete="$emit('onCellEditComplete', $event)"
        editMode="cell"
      >
        <Column
          v-for="{ field, header, type } in table.headers"
          :sortable="sortableColumn"
          :key="field"
          :field
          :header="$t(header)"
        >
          <template #body="{ data, field }">
            <slot :name="field" :value="path(field.split('.'), data)" :type="type">
              {{ isNil(type) ? path(field.split('.'), data) || '-' : getFormattedData(data, field, type) }}
            </slot>
          </template>

          <template #editor="{ data, field }" v-if="isFieldEditable(notEditableColumns, field)">
            <template v-if="calendarFields.includes(field)">
              <Calendar v-model="data[field]" dateFormat="dd.mm.yy" />
            </template>
            <template v-else-if="currencyFields.includes(field)">
              <UiNumberInput v-model="data[field]" mode="currency" currency="USD" />
            </template>
            <template v-else-if="numberFields.includes(field)">
              <UiNumberInput v-model="data[field]" autofocus />
            </template>

            <template v-else>
              <UiInput v-model="data[field]" />
            </template>
          </template>
        </Column>

        <template #empty>{{ $t('table.empty') }}</template>
      </DataTable>
    </template>
  </Card>
</template>

<style scoped></style>
