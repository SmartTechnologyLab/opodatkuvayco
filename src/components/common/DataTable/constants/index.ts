import { _t } from '@/i18n'
import type { TableHeaders } from '../types'
import { Currency } from '@/constants/currencies'

export enum FormatType {
  Date,
  Number,
  CurrencyUAH,
  CurrencyUSD,
  CurrencyEUR,
  Percent
}

export const dynamicCurrencies = [FormatType.CurrencyUSD, FormatType.CurrencyEUR]

export const currenciesName: Record<Currency, Currency> = {
  [Currency.UAH]: Currency.UAH,
  [Currency.USD]: Currency.USD,
  [Currency.EUR]: Currency.EUR
}

export const resultHeaders: TableHeaders[] = [
  {
    header: _t('table.headers.ticker'),
    field: 'ticker'
  },
  {
    header: _t('table.headers.purchase-date'),
    field: 'purchase.date',
    type: FormatType.Date
  },
  {
    header: _t('table.headers.quantity'),
    field: 'quantity',
    type: FormatType.Number
  },
  {
    header: _t('table.headers.purchase'),
    field: 'purchase.price',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.sum'),
    field: 'purchase.sum',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.comission'),
    field: 'purchase.commission',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.rate'),
    field: 'purchase.rate',
    type: FormatType.CurrencyUAH
  },
  {
    header: _t('table.headers.purchase'),
    field: 'purchase.uah',
    type: FormatType.CurrencyUAH
  },
  {
    header: _t('table.headers.sale-date'),
    field: 'sale.date',
    type: FormatType.Date
  },
  {
    header: _t('table.headers.price'),
    field: 'sale.price',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.sum'),
    field: 'sale.sum',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.comission'),
    field: 'sale.commission',
    type: FormatType.CurrencyUSD
  },
  {
    header: _t('table.headers.rate'),
    field: 'sale.rate',
    type: FormatType.CurrencyUAH
  },
  {
    header: _t('table.headers.sale'),
    field: 'sale.uah',
    type: FormatType.CurrencyUAH
  },
  {
    header: _t('table.headers.total'),
    field: 'total',
    type: FormatType.CurrencyUAH
  },
  {
    header: _t('table.headers.percent'),
    field: 'percent',
    type: FormatType.Percent
  }
]
