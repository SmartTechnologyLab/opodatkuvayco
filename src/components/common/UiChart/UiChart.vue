<script setup lang="ts">
import { computed, toRefs, useAttrs } from 'vue'
import Chart, { type ChartProps } from 'primevue/chart'
import type { ChartData } from './common/types'
import { mergeRight } from 'ramda'
import { DEFAULT_CHART_OPTIONS } from './common/constants'

const attrs = useAttrs()

const props = defineProps<{
  chartData: ChartData
  options?: ChartProps['options']
}>()

const { options } = toRefs(props)

const mergedOptions = computed(() => mergeRight(DEFAULT_CHART_OPTIONS, options.value))
</script>

<template>
  <Chart :data="chartData" :options="mergedOptions" v-bind="{ ...attrs }" />
</template>

<style scoped lang="scss"></style>
