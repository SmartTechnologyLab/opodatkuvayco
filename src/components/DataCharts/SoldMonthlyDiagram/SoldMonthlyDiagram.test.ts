import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { shallowMount, type VueWrapper } from '@vue/test-utils'
import SoldMonthlyGraph from '@/components/DataCharts/SoldMonthlyDiagram/SoldMonthlyDiagram.vue'
import { type Deal, type Operation } from '@/components/common/DataTable/mocks'
import UiChart from '@/components/common/UiChart/UiChart.vue'
import { nextTick } from 'vue'
import { stackedGraphOptions } from '@/components/DataCharts/common/constants'
import UiCard from '@/components/common/UiCard/UiCard.vue'
import { stubComponent } from '@/helpers/testHelpers/createComponent'

vi.mock('@/components/DataCharts/common/helpers', () => ({
  getResultByTickerEachMonth: vi.fn().mockImplementation((operation: Operation) => {
    return () => {
      if (operation === 'sale') {
        return {
          labels: ['січень'],
          datasets: [
            {
              label: 'MSFT',
              data: [23166, ...Array(11).fill(0)]
            }
          ]
        }
      } else {
        return {
          labels: ['січень'],
          datasets: [
            {
              label: 'MSFT',
              data: [7321.86, ...Array(11).fill(0)]
            }
          ]
        }
      }
    }
  })
}))

const mockDeals: Deal[] = [
  {
    id: Symbol(),
    purchase: {
      commission: 2.7,
      date: new Date('2021-01-10'),
      price: 15,
      rate: 26,
      sum: 270,
      uah: 7321.86
    },
    quantity: 18,
    sale: {
      commission: 5.94,
      date: new Date('2021-01-11'),
      price: 33,
      rate: 39,
      sum: 594,
      uah: 23166
    },
    ticker: 'MSFT',
    total: 15844.14,
    percent: 2.163950143815916
  }
]

const UiCardStub = stubComponent(UiCard, {
  template: `
    <div>
      <slot name="title"></slot>
      <slot />
    </div>
  `
})

describe('SoldMonthlyDiagram', () => {
  let wrapper: VueWrapper<InstanceType<typeof SoldMonthlyGraph>>

  const getCardTitle = (operation: 'sale' | 'purchase') =>
    `Діаграма ${operation === 'purchase' ? 'придбань' : 'продажів'} за рік по активам`

  beforeEach(() => {
    wrapper = shallowMount(SoldMonthlyGraph, {
      global: {
        stubs: {
          UiChart,
          UiCard: UiCardStub
        }
      },
      props: {
        deals: mockDeals,
        stacked: false,
        operation: 'sale'
      }
    })
  })

  afterEach(() => {
    wrapper.unmount()

    vi.clearAllMocks()
  })

  it('should contain type of bar', () => {
    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.attributes().type).toStrictEqual('bar')
  })

  it('should group result when operation is sale', () => {
    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().chartData).toMatchSnapshot()
  })

  it('should group result when operation is purchase', async () => {
    await wrapper.setProps({ operation: 'purchase' })

    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().chartData).toMatchSnapshot()
  })

  it('should contain custom chart options when props stacked is passed', async () => {
    wrapper.setProps({ stacked: true })

    await nextTick()

    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().options).toEqual(stackedGraphOptions)
  })

  it('should not contain custom options when stacked is not passed', async () => {
    wrapper.setProps({ stacked: false })

    await nextTick()

    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().options).toBeUndefined()
  })

  it('should render sell title', () => {
    const CARD_TITLE = getCardTitle(wrapper.props().operation)

    const uiCard = wrapper.findComponent(UiCard)

    expect(uiCard.text()).toContain(CARD_TITLE)
  })

  it('should render purchase title', async () => {
    await wrapper.setProps({ operation: 'purchase' })

    const CARD_TITLE = getCardTitle(wrapper.props().operation)

    const uiCard = wrapper.findComponent(UiCard)

    expect(uiCard.text()).toContain(CARD_TITLE)
  })
})
