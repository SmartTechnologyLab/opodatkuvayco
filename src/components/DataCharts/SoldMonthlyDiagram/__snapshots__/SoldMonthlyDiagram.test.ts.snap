// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SoldMonthlyDiagram > should group result when operation is purchase 1`] = `
{
  "datasets": [
    {
      "data": [
        7321.86,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "label": "MSFT",
    },
  ],
  "labels": [
    "січень",
  ],
}
`;

exports[`SoldMonthlyDiagram > should group result when operation is sale 1`] = `
{
  "datasets": [
    {
      "data": [
        23166,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "label": "MSFT",
    },
  ],
  "labels": [
    "січень",
  ],
}
`;
